File: go/examples/otel_example.go
package main

import (
	"context"
	"fmt"
	"log"
	"time"

	"github.com/atm/atm/pkg/utils/otel"
	"go.opentelemetry.io/otel/attribute"
)

func main() {
	ctx := context.Background()

	// Initialize OpenTelemetry
	// Can be controlled via environment variables:
	//   export ATM_OTEL_ENABLED=true
	//   export ATM_OTEL_ENDPOINT=localhost:4317
	//   export ATM_SERVICE_NAME=atm-example

	if err := otel.Initialize(ctx, nil); err != nil {
		log.Fatalf("Failed to initialize OpenTelemetry: %v", err)
	}
	defer otel.Shutdown(ctx)

	// Example 1: Using TraceFunc helper
	err := otel.TraceFunc(ctx, "example.operation", func(ctx context.Context) error {
		fmt.Println("Executing operation...")
		time.Sleep(100 * time.Millisecond)
		return nil
	}, attribute.String("component", "example"))
	if err != nil {
		log.Printf("Operation failed: %v", err)
	}

	// Example 2: Manual tracing
	tracer := otel.GetTracer("example")
	ctx, span := tracer.Start(ctx, "manual.operation", attribute.String("type", "manual"))
	defer span.End()

	fmt.Println("Inside manual trace")
	time.Sleep(50 * time.Millisecond)

	// Example 3: Metrics - Counter
	counter := otel.NewCounter("example.operations.total", "Total operations")
	counter.Add(ctx, 1, attribute.String("status", "success"))
	counter.Add(ctx, 1, attribute.String("status", "error"))

	// Example 4: Metrics - Histogram
	histogram := otel.NewHistogram("example.duration", "Operation duration", "ms")
	histogram.Record(ctx, 150.5, attribute.String("operation", "data_sync"))
	histogram.Record(ctx, 200.3, attribute.String("operation", "data_sync"))

	// Example 5: Record duration
	err = otel.RecordDuration(ctx, "example.duration", func(ctx context.Context) error {
		time.Sleep(100 * time.Millisecond)
		return nil
	}, attribute.String("component", "example"))
	if err != nil {
		log.Printf("Duration recording failed: %v", err)
	}

	fmt.Println("OpenTelemetry example completed")
}

File: python/examples/otel_example.py
"""
Example: Using OpenTelemetry in ATM project.

Demonstrates how to use OpenTelemetry for logs, metrics, and traces.
"""

import logging
import os
import time

from atm.utils.otel import initialize_otel, get_tracer, get_meter
from atm.utils.otel_helpers import (
    MetricsCounter,
    MetricsHistogram,
    record_duration,
    trace_function,
    trace_span,
)
from atm.utils.otel_logger import setup_logger

# Configure logging
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)


@trace_function(attributes={"component": "example", "operation": "demo"})
@record_duration("example.duration", attributes={"component": "example"})
def example_function():
    """Example function with tracing and metrics."""
    logger.info("Executing example function")
    time.sleep(0.1)  # Simulate work
    return "success"


def main():
    """Run OpenTelemetry example."""
    # Initialize OpenTelemetry
    # Can be controlled via environment variables:
    #   export ATM_OTEL_ENABLED=true
    #   export ATM_OTEL_ENDPOINT=localhost:4317
    #   export ATM_SERVICE_NAME=atm-example

    initialize_otel(
        service_name="atm-example",
        service_version="0.1.0",
        endpoint=os.getenv("ATM_OTEL_ENDPOINT", "localhost:4317"),
        enabled=os.getenv("ATM_OTEL_ENABLED", "false").lower() == "true",
    )

    # Setup logger (format stays the same)
    logger = setup_logger("example")
    logger.info("OpenTelemetry example started")

    # Example 1: Using decorators
    logger.info("Example 1: Using decorators")
    result = example_function()
    logger.info(f"Result: {result}")

    # Example 2: Using context manager
    logger.info("Example 2: Using context manager")
    with trace_span("manual_operation", attributes={"type": "manual"}):
        logger.info("Inside trace span")
        time.sleep(0.05)

    # Example 3: Manual tracing
    logger.info("Example 3: Manual tracing")
    tracer = get_tracer(__name__)
    if tracer:
        with tracer.start_as_current_span("manual_trace") as span:
            span.set_attribute("custom.attribute", "value")
            logger.info("Inside manual trace")
            time.sleep(0.05)

    # Example 4: Metrics - Counter
    logger.info("Example 4: Metrics - Counter")
    counter = MetricsCounter("example.operations.total", description="Total operations")
    counter.add(1, attributes={"status": "success"})
    counter.add(1, attributes={"status": "error"})

    # Example 5: Metrics - Histogram
    logger.info("Example 5: Metrics - Histogram")
    histogram = MetricsHistogram("example.duration", description="Operation duration", unit="ms")
    histogram.record(150.5, attributes={"operation": "data_sync"})
    histogram.record(200.3, attributes={"operation": "data_sync"})

    # Example 6: Using meter directly
    logger.info("Example 6: Using meter directly")
    meter = get_meter(__name__)
    if meter:
        counter = meter.create_counter("example.custom.counter", description="Custom counter")
        counter.add(1, attributes={"source": "example"})

    logger.info("OpenTelemetry example completed")


if __name__ == "__main__":
    main()

